components/service/glean/src/test/java/mozilla/components/service/glean/StringListMetricTypeTest.kt

```diff
@@ -0,0 +1,200 @@
+import kotlinx.coroutines.ObsoleteCoroutinesApi
+import mozilla.components.service.glean.StringListMetricType.Companion.MAX_STRING_LENGTH
+import mozilla.components.service.glean.storages.StringListsStorageEngine
+import mozilla.components.service.glean.storages.StringListsStorageEngineImplementation
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNull
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+
+@ObsoleteCoroutinesApi
+@ExperimentalCoroutinesApi
+@RunWith(RobolectricTestRunner::class)
+class StringListMetricTypeTest {
+
+    @get:Rule
+    val fakeDispatchers = FakeDispatchersInTest()
+
+    @Before
+    fun setUp() {
+        StringListsStorageEngine.applicationContext = ApplicationProvider.getApplicationContext()
+        // Clear the stored "user" preferences between tests.
+        ApplicationProvider.getApplicationContext<Context>()
+            .getSharedPreferences(
+                StringListsStorageEngine.javaClass.simpleName,
+                Context.MODE_PRIVATE)
+            .edit()
+            .clear()
+            .apply()
+        StringListsStorageEngine.clearAllStores()
+    }
+
+    @Test
+    fun `The API must define the expected "default" storage`() {
+        // Define a 'stringMetric' string metric, which will be stored in "store1"
+        val stringListMetric = StringListMetricType(
+            disabled = false,
+            category = "telemetry",
+            lifetime = Lifetime.Ping,
+            name = "string_list_metric",
+            sendInPings = listOf("store1")
+        )
+        assertEquals(listOf("metrics"), stringListMetric.defaultStorageDestinations)
+    }
+
+    @Test
+    fun `The API saves to its storage engine`() {
+        // Define a 'stringMetric' string metric, which will be stored in "store1"
+        val stringListMetric = StringListMetricType(
+            disabled = false,
+            category = "telemetry",
+            lifetime = Lifetime.Application,
+            name = "string_list_metric",
+            sendInPings = listOf("store1")
+        )
+
+        // Record two lists using add and set
+        stringListMetric.add("value1")
+        stringListMetric.add("value2")
+        stringListMetric.add("value3")
+
+        // Check that data was properly recorded.
+        val snapshot = StringListsStorageEngine.getSnapshot("store1", false)
+        assertEquals(1, snapshot!!.size)
+        assertEquals(true, snapshot.containsKey("telemetry.string_list_metric"))
+        assertEquals("value1", snapshot.get("telemetry.string_list_metric")?.get(0))
+        assertEquals("value2", snapshot.get("telemetry.string_list_metric")?.get(1))
+        assertEquals("value3", snapshot.get("telemetry.string_list_metric")?.get(2))
+
+        // Use set() to see that the first list is replaced by the new list
+        stringListMetric.set(listOf("other1", "other2", "other3"))
```

Comment:
If we move the `set` before the previous `add`(s) then we can also make sure that adding correctly appends to the current list of values. It would also simplify the test by having one single snapshot :)

Reply:
This doesn't test that set replaces the current list though.  The second and third `add` should test appending to the current list, I also wanted to see that `set` actually replaced the value, similar to `set` in the String metric type.

Reply:
This is a good point! Can we then have another test that sets the list, then adds to it, and then makes sure that the returned data comes from both the set and the add?

Reply:
Added a test that does `set` first, followed by `add` and named the old test and the new test appropriately